control_protocol.md

```diff
@@ -0,0 +1,293 @@
+
+{ref}`Components <components.md#components>` shall have one DEALER socket connecting to one Coordinator's ROUTER socket.
+
+Coordinators shall have one DEALER socket per other Coordinator in the Network.
+This DEALER socket shall connect to the other Coordinator's ROUTER socket.
+
+Messages must be sent to a Coordinator's ROUTER socket.
+
+
+(router-sockets)=
+#### Particularities of ROUTER sockets
+
+A small introduction to ROUTER sockets, for more details see [zmq guide chapter 3](https://zguide.zeromq.org/docs/chapter3/#Exploring-ROUTER-Sockets).
+
+A router socket assigns a random _identity_ to each connecting peer.
+If, for example, two Components `CA`, `CB` connect to the ROUTER sockets, the socket assigns the identities, which we call `IA`, `IB` (they can be any bytes sequence).
+
+Whenever a message is sent to the ROUTER socket from a peer, it reads as a first frame that identity.
+For example if `CA` sends the message `Request A`, the ROUTER socket will read `IA|Request A`.
+That way, you can return an answer to exactly the original peer and not, for example `CB`.
+If you call the ROUTER's send command with `IA|Reply A`, the socket will send `Reply A` to the peer, whose connection is `IA`, in this case `CA`.
+
+The following diagram shows this example communication with two Components:
+:::{mermaid}
+graph TD
+    CA([CA]) -->|"Request A"| D[recv] -->|"IA|Request A"| A([Coordinator])
+    CB([CB]) -->|"Request B"| C[recv] -->|"IB|Request B"| A
+    subgraph ROUTER socket
+        B
+        C
+        D
+        E
+    end
+    A -. "IA|Reply A" .-> B
+    B[send] -. "Reply A" .-> CA
+    A -. "IB|Reply B" .-> E
+    E[send] -. "Reply B" .-> CB
+:::
+
+
+### Protocol basics
+
+#### Naming scheme
+
+Each Component has an individual ID, given by the user, the _Component ID_.
+A Component ID must be a series of bytes, without the ASCII character "." (byte value 46).
+Component IDs must be unique in a {ref}`Node <network-structure.md#node>`, i.e. among the Componets connected to a single Coordinator.
+The Coordinator itself has the Component ID `COORDINATOR`.
+:::{note}
+COORDINATOR is a placeholder for the final version.
+:::
+
+Also every Node ID has to be unique in the Network.
+As each Component belongs to exactly one Coordinator, it is fully identified by the combination of Node ID and Component ID., which is globally unique.
+This _full ID_ is the composition of Node ID, ".", and Component ID.
+For example `Co1.CA` is the full ID of the Component `CA` in the Node `Co1`.
+
+The receiver of a message may be specified without the Node ID, if the receiver lives in the same Node.
+
+
+#### Message composition
+
+A message consists in two or more frames.
+1. The receiver full ID.
+2. The sender full ID.
+3. Message content: The optional payload, which can be 0 or more frames.
+
+
+(address-book)=
+#### Address book
```

Comment:
A refinement of terms: 
We should use different terms for different things, so it's easier to get a clear picture. What above is an "ID" is here an "Address" (iiuc); further above the random zmq designator is called "identity" (the abbrevation of which is generally "ID").

Can we/should we get a cleaner separation here, e.g. by calling the user-chosen designator "address" (local and global) and the random zmq thing "ID"? Or some other combination (local "name" and global "address", vs. "ID")?

Reply:
A third related/overloaded term (see e.g. https://github.com/pymeasure/leco-protocol/issues/22#issuecomment-1414403570) is the "address" meaning the host:port combination that zmq sockets live at.

Ideally, to avoid confusing ourselves and others, we find/choose different (compound) words for these different concepts, and stick to them.

Reply:
I used address for host / port (TCP), except in address book 🤦‍♂️; ID for our own naming scheme; (connection) identity for the Zmq thing, as it is thir term.

Could we use "Component name" and Node name?

Reply:
Address is given by popular usage, identity is the definition of zmq. So they are both fixed.

Reply:
I'm not sure, whether "address book" is a good name, even though it just contains names, because it is similar to a normal address book.

Or should we call it Directory?

Reply:
"Directory" sounds good to me!

> I used address for host / port (TCP)
> (connection) identity for the Zmq thing, as it is their term.

Agreed to both.

> ID for our own naming scheme; 

I don't like how we collide because "ID" is just the abbrevation for "identity"/identification, this is bound to confuse.

> Could we use "Component name" and Node name?

We could use the parts "Namespace" and "(Component) name"; the first we have already established. The second may also be "obviously" clear, and we could probably use "Name" alone. 
How do we call the whole thing? "Path" maybe? (not 100% convinced)

Reply:
"Full name" for the combination?

Path reminds me of file system, but on the other hand, we have a similar syntax: `path = folder.file`

Reply:
👍 to Full name.

Reply:
full name is good for me too